kernel/network/e1000.c

```diff
@@ -60,31 +64,126 @@ read_mac()
 
-  writeCommand(REG_RXDESCLO, (uint32_t)e1000.rx_descs );
-  writeCommand(REG_RXDESCHI, 0);
+}
+
+void
+init_transmit_descriptors()
+{
+  write_command(REG_TXDESCLO, (uint32_t) e1000->tx_descs);
+  write_command(REG_TXDESCHI, 0);
 
-  // set len to number of descriptors * sizeof one in byte
-  writeCommand(REG_RXDESCLEN, E1000_NUM_RX_DESC * 16);
+  //now setup total length of descriptors
 
-  writeCommand(REG_RXDESCHEAD, 0);
-  writeCommand(REG_RXDESCTAIL, E1000_NUM_RX_DESC-1);
-  e1000.rx_cur = 0;
-  writeCommand(REG_RCTRL, RCTL_EN | RCTL_SBP| RCTL_UPE | RCTL_MPE | RCTL_LBM_NONE | RTCL_RDMTS_HALF | RCTL_BAM | RCTL_SECRC  | RCTL_BSIZE_8192);
+  write_command(REG_TXDESCLEN, E1000_NUM_TX_DESC * sizeof(e1000_tx_desc_t));
+
+  //setup numbers
+  write_command(REG_TXDESCHEAD, 0);
+  write_command(REG_TXDESCTAIL, 0);
+  e1000->tx_cur = 0;
+  // works with e1000 only
+  write_command(REG_TCTRL, read_command(REG_TCTRL) | TCTL_EN | TCTL_PSP);
+  write_command(REG_TIPG,  0x0060200A);
+}
+
+void
+enable_interrupt()
+{
+  // according to E1000 documentation p. 297
+  write_command(REG_IMASK, 0x1F2D4);
+}
+
+void start_link(void)
+{
+  uint32_t flags = read_command(REG_CTRL);
+  write_command(REG_CTRL, flags | ECTRL_SLU);
+}
 
+void
+irq_handler_e1000(void)
+{
+  // clear interrupt on device
+  write_command(REG_IMASK, 0x1);
+  uint32_t cause = read_command(REG_INT_CAUSE);
+  if (cause & LSC)
+    start_link();
+  if (cause & RXDMT0)
+    ; // good threshold
+  if (cause & RXT0)
+    receive_packet();
 }
 
 void
-e1000_start(void)
+start_e1000(void)
 {
   detect_eeprom();
   read_mac();
+  start_link();
+  enable_interrupt();
+  init_receive_descriptors();
+  init_transmit_descriptors();
+}
+
+int
+send_packet(const void *p_data, uint16_t p_len)
+{
+  printf("[E1000] Sending packet data <%s> with len %i\n", p_data, p_len);
+  e1000_tx_desc_t *curr = &(e1000->tx_descs[e1000->tx_cur]);
+  curr->addr = (uint32_t) p_data;
+  curr->length = p_len;
+  curr->cmd = CMD_EOP | CMD_IFCS | CMD_RS;
+  curr->status = 0;
+  uint8_t old_cur = e1000->tx_cur;
+  e1000->tx_cur = (e1000->tx_cur + 1) % E1000_NUM_TX_DESC;
+
+  write_command(REG_TXDESCTAIL, e1000->tx_cur);
+
+  while(!(curr->status));
+  printf("[E1000] Packet send!\n"); 
+  return 0;
+}
+
+void
+receive_packet()
+{
+  uint16_t old_cur;
+  uint8_t got_packet = 0;
+
+  while(e1000->rx_descs[e1000->rx_cur].status & 0x1)
+    {
+      got_packet = 1;
+      uint8_t* buf = (uint8_t *) ((uint32_t) e1000->rx_descs[e1000->rx_cur].addr);
+      uint16_t len = e1000->rx_descs[e1000->rx_cur].length;
+
+      log_debug("received packet with length: %i", len)
+
+      e1000->rx_descs[e1000->rx_cur].status = 0;
+      old_cur = e1000->rx_cur;
+      e1000->rx_cur = (e1000->rx_cur + 1) % E1000_NUM_RX_DESC;
+      write_command(REG_RXDESCTAIL, old_cur);
+    }
 }
 
 void
-e1000_init(void)
+init_e1000(void)
```

Comment:
why do we have `init_e1000` and `start_e1000`? Couldn't we call a method like `get_memory` that is called in `start_e1000` or something like that?

Reply:
Seems reasonable, but I think for the sake of extendability and maybe in the future having support for multiple network cards (maybe even at the same time lol) we want to have some method that can startup or shutdown (that is not present yet) our network interface just like `ifup` does